Refactor EngineFunctions / GCS / IO and enable cross FS localization

Horneth opened the pull request:
This is an attempt to make the separation between (JES / Local / SGE) and File systems (GCS, SharedFilesystem) more obvious. It enables implementation of more Backend / File systems combinations.
For example this wdl runs, on a Local cromwell instance :
Wdl: 
```
task hello {
  String addressee
  command {
    echo "Hello ${addressee}!"
  }
  output {
    String fileoutput = read_string(stdout())
    String salutation = read_string("gs://tjeandet/wdl_input_public.txt")
  }
}

workflow hello {
  File infile
  String instring = read_string(infile)
  call hello {input: addressee = instring}
}
```

Inputs:
```
{
    "hello.infile": "gs://tjeandet/wdl_input_public.txt"
}
```
It also works with user-defined authentication for GCS (refresh token).
More tests / cleaning is needed before it's reviewable but I wanted to put it out there before Thanksgiving.

mcovarr commented on src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala:
```diff
@@ -1,17 +1,18 @@
 
   /**
-    * This method should try its best to provide a GCS connection setup with the user's credentials.
+    * This method should try its best to provide a GCS connection setup with the user's crendentials.
```
crendentials -> credentials

scottfrazer commented:
Wow, this will be really awesome!

cjllanwarne commented:
Just as a safety measure, maybe we should consider add a config option called something like "allowCrossFSLocalisation" - just to avoid accidentally downloading a 500GB file from GCS and the costs involved?

Horneth commented:
Yeah I thought about that too. A config flag might be a good safety net.

geoffjentry commented on src/main/scala/cromwell/binding/expression/WdlFunctions.scala:
```diff
@@ -16,16 +16,14 @@ trait WdlFunctions[T] {
-  def fileContentsToString(path: String): String = throw new NotImplementedError("fileContentsToString() is unimplemented")
+    * Path where to write files created by engine functions (write_*).
+   */
+  def tempFilePath: String = throw new NotImplementedError("tempFilePath() is unimplemented")
```
Doesn't `???` throw a `NotImplementedError`?

cjllanwarne replied:
But when I read it, ??? looks like a temporary placeholder whereas this looks like a conscious decision.

geoffjentry replied:
I'd change the message then to make it clear that it was a conscious decision. To me, saying "x is unimplemented" as the message for a NotImplementedError seems redundant vs. saying something like "is not supported" (just as an example)

cjllanwarne replied:
I'd agree with that

geoffjentry replied:
You better!!!!!!!!

geoffjentry commented on src/main/scala/cromwell/engine/backend/jes/JesConfiguration.scala:
```diff
@@ -1,16 +1,26 @@
 
 object ProductionJesConfiguration {
   lazy val jesConf =  JesAttributes()
+  lazy val googleConf =  GoogleConfiguration.gcloudConf.get
```
what happens if the get fails?

Horneth replied:
`GoogleConfiguration.gcloudConf` is a `Try` wrapping the parsing of the Google configuration, which is using scalaz to validate the configuration.
So it fails an exception containing all the reasons while the config is not correct will be thrown.

geoffjentry commented on src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala:
```diff
@@ -1,30 +1,33 @@
+  lazy val jesCromwellInterface: JesInterface = {
+    val cromwellCredentials = GoogleCredentialFactory.fromCromwellAuthScheme
+    // .get to fail now, as we can't run on Jes without a Cromwell authenticated GCS interface
+    val gcsInterface = GoogleCloudStorage.cromwellAuthenticated.get
```
You could have it throw a nicer error though via a getOrElse(throw new ....)

Horneth replied:
Agreed

Horneth replied:
Oh actually this one is a `Try` which encapsulates the potential configuration parsing errors leading to the failure (same as your other comment).
However the one line 70 for the `jesUserConnection` is an Option and definitely could use a better exception.

geoffjentry commented on src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala:
```diff
@@ -1,98 +1,21 @@
-    } yield file
-  }
+  override def fileContentsToString(path: String) = {
+    if(!Paths.get(path).isAbsolute && !path.isUriWithProtocol)
```
space after if

geoffjentry commented on src/main/scala/cromwell/engine/io/gcs/GoogleConfiguration.scala:
```diff
@@ -0,0 +1,70 @@
+
+// Google Authentication modes supported for Cromwell
+sealed trait GoogleCromwellAuthMode
+case class ServiceAccountMode(accountId: String, pemPath: String) extends GoogleCromwellAuthMode
```
make these case classes final

geoffjentry commented on src/main/scala/cromwell/engine/io/shared/SharedFileSystemIOInterface.scala:
```diff
@@ -0,0 +1,42 @@
+
+import scala.util.Try
+
+class SharedFileSystemIOInterface extends IOInterface {
```
You're not being consistent with IO. Down below you have `ioManager` treating the io as a single word and here IOInterface as two. I prefer the former (i.e. SharedFileSystemIoInterface) but choose one and stick to it

geoffjentry commented on src/main/scala/cromwell/engine/io/shared/SharedFileSystemIOInterface.scala:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  override def glob(path: String, pattern: String): Seq[String] = {
+    Paths.get(path).glob(s"**/$pattern") map {
```
This map looks like it'd fit on one line

geoffjentry commented:
@cjllanwarne @Horneth re the saving people from themselves, while I don't completely disagree what would be the use case here? If someone is running in local or SGE mode they'd need to be running Cromwell with (for now) Google credentials which means they probably know what they're doing.

The "don't completely disagree" part is that I support the idea in general but think it should default to the unsafe direction as IMO it's more likely someone gets bit by it being turned off than turned on.

mcovarr commented on src/main/scala/cromwell/engine/io/IOManager.scala:
```diff
@@ -0,0 +1,38 @@
+class IoManager(interfaces: Seq[IoInterface]) extends IoInterface {
+
+  private def findInterface(path: String) = interfaces.find(_.isValidPath(path))
+  private def unsupported(path: String) = throw new UnsupportedOperationException(s"No IOInterface able to process $path has been found.")
```
wrong IoException case in exception text

mcovarr commented on src/main/scala/cromwell/engine/io/IOManager.scala:
```diff
@@ -0,0 +1,38 @@
+
+import cromwell.binding.IoInterface
+
+//TODO: Use reflection to automagically implement all IOInterface methods ?
```
Not sure about reflection, but there might be some possibilities with delegation/AOP macros.  Or at least this might be DRYer with the `findInterface(path) map { _.fn } getOrElse unsupported(path)` pulled out to a function.

Horneth replied:
But how do you know what `fn` is ? Also it's going to need parameters :/

mcovarr replied:
```scala
  
  private def withInterface[T](path: String, fn: IoInterface => T): T = {
    findInterface(path) map fn getOrElse unsupported(path)
  } 

  override def writeFile(path: String, content: String) = withInterface(path, _.writeFile(path, content))
```

Horneth replied:
Ah ok I didn't think of it that way, nice.

geoffjentry replied:
I noticed the words reflection and macros when scanning my email this morning. If either of those end up being the final proposed solution, let's chat :)

mcovarr commented on src/main/scala/cromwell/util/PathUtil.scala:
```diff
@@ -0,0 +1,11 @@
+
+  implicit class UriString(val str: String) extends AnyVal {
+    def isGcsUrl: Boolean = str.startsWith("gs://")
+    def isUriWithProtocol: Boolean = "^[a-zA-Z]+://".r.findFirstIn(str).nonEmpty
```
are there capital letters in protocols?

Horneth replied:
Yeah probably not

mcovarr commented on src/main/scala/cromwell/binding/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -1,14 +1,25 @@
 
 trait WdlStandardLibraryFunctions extends WdlFunctions[WdlValue] {
+
+  def interface: IoInterface = throw new NotImplementedError("No IOInterface was implemented")
```
wrong case in exception text

mcovarr commented on src/main/scala/cromwell/binding/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -1,14 +1,25 @@
+
+  def interface: IoInterface = throw new NotImplementedError("No IOInterface was implemented")
+
   private def fail(name: String) = Failure(new UnsupportedOperationException(s"$name() not implemented yet"))
```
sounds more like a `NotImplementedError`

mcovarr commented on src/main/scala/cromwell/engine/io/IOManager.scala:
```diff
@@ -0,0 +1,38 @@
+package cromwell.engine.io
```
File name should be `IoManager.scala`

mcovarr commented on src/main/scala/cromwell/engine/io/IOManager.scala:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  override def writeFile(path: String, content: String): Unit = {
+    findInterface(path) map { _.writeFile(path, content) } getOrElse unsupported(path)
```
IntelliJ warns for map with Unit return type

geoffjentry replied:
only as of a year or two ago :'(

mcovarr commented:
Per the comments above is this meant to be a "stealth" feature or should there be something in the README.md to note its existence, config requirements, gotchas?

mcovarr commented on src/main/scala/cromwell/binding/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -91,16 +102,66 @@ trait WdlStandardLibraryFunctions extends WdlFunctions[WdlValue] {
+      tsvSerialized <- singleArgument.asInstanceOf[WdlArray].tsvSerialize
+      file <- writeContent("array", tsvSerialized)
+    } yield file
+  }
```
Could the above be refactored to be more DRY?

mcovarr commented:
We *really* need to have an environment for automated testing of GCS / JES stuff.  :cry:

mcovarr commented on src/main/scala/cromwell/engine/EngineFunctions.scala:
```diff
@@ -0,0 +1,28 @@
+
+/**
+ * Default implementation of Wdl Standard Library functions executed at the workflow level.
+ * This class is abstract to enforce that Backends extend it and customize the behaviour if needed.
```
behaviour -> behavior
:us:

mcovarr commented:
:+1: delta some minorish comments

scottfrazer commented on src/main/scala/cromwell/engine/EngineFunctions.scala:
```diff
@@ -0,0 +1,28 @@
+abstract class WorkflowEngineFunctions(override val interface: IoInterface, val context: WorkflowContext) extends WdlStandardLibraryFunctions {
+
+  override def tempFilePath: String = context.root
+  override def globPath(glob: String): String = context.root
```
I don't think I understand what `globPath` is supposed to be doing here

mcovarr replied:
same here

Horneth replied:
Yeah I'm not a fan of this either. The idea was to 'factorize' the "glob" function like the others in `WdlStandardLibraryFunctions`, because it's almost always the same code. In short that's a way to tell the `glob` function in `WdlStandardLibraryFunctions` where to glob from, which differs from backend to backend. But it's kind of hacky, I'll see if I can find another way or just have each backend implementation override this with its own implementation.

scottfrazer replied:
Can you maybe give an example of one case in which it's used?  I still have a hard time following it

Horneth replied:
So what happens is "how to glob" is independent from the backend, it only depends on the file system (Shared, GCS). This is good because it allows to have a `glob` method in `IoInterface`, that each file system implements to say how to glob some files out from a directory path.

However "where to glob from" is different depending on the backend. For instance local backend doesn't have anything specific regarding globbing, so you can just apply the pattern to the call directory. But JES delocalizes all the files into a dedicated directory inside the call directory. And this is where you need to go and look for the files.

There are different way to do it but in the end what's needed is a way for each backend to tell the `glob` function in `WdlStandardLibraryFunctions` where to glob from, and this the what this `globPath` method is for. It's only used in the `glob` function implementation in `WdlStandardLibraryFunctions`. And it's overridden in `JesWorkflowEngineFunctions`:

```
override def globPath(glob: String) = s"${context.root}/glob-${glob.md5Sum}/"
```

The more general idea is to minimize the need to reimplement engine functions for each backend, and to do that we need to identify the smaller set of things needed to implement them that are backend specific, so Backends only have those to implement.
The advantage of this is that if you look at 
`JesWorkflowEngineFunctions`, `JesCallEngineFunctions`, `LocalWorkflowEngineFunctions`and  `LocalCallEngineFunctions` they're almost empty, because everything is already in `WdlStandardLibraryFunctions`.

scottfrazer replied:
I see.  Thanks for the detailed explanation, I forgot to account for the fact that JES has to delocalize the glob files into its own directory

mcovarr commented on src/main/scala/cromwell/binding/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -1,13 +1,23 @@
 trait WdlStandardLibraryFunctions extends WdlFunctions[WdlValue] {
-  private def fail(name: String) = Failure(new UnsupportedOperationException(s"$name() not implemented yet"))
+
+  def interface: IoInterface = throw new NotImplementedError("No IoInterface was implemented")
```
can this just be 

```scala
def interface: IoInterface
```

to force the issue at compile time?

Horneth replied:
Yep

mcovarr commented:
:+1: again delta even fewer and more minor comments

scottfrazer commented on src/main/scala/cromwell/binding/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -91,16 +101,55 @@ trait WdlStandardLibraryFunctions extends WdlFunctions[WdlValue] {
+  override def writeTempFile(path: String, prefix: String, suffix: String, content: String): String = throw new NotImplementedError()
+  override def exists(path: String): Boolean = throw new NotImplementedError()
+  override def interface: IoInterface = throw new NotImplementedError()
+  override def isValidPath(path: String): Boolean = throw new NotImplementedError()
```
I can't remember exactly... Is `???` better to use than `throw new NotImplementedError()`?

geoffjentry replied:
Personally I view them as the same although in the last few days multiple folks have expressed that ??? seems like more of a placeholder than an intentional thing. That said, if one's going to have those exceptions, to go along with the making it explicit angle I'd like to have a message along the lines of "foo is not currently supported".

scottfrazer commented:
Besides the glob thing, I give it a :+1:

Horneth commented:
Just wanted to re touch base on this quickly, since the :+1: were given I fixed a bug that prevented the "filepassing.wdl" from working on JES and updated the config template / README.

Is updating the template enough to make any configuration change work when cromwell will redeploy ? Because several field names / location have been changed.

mcovarr commented:
Is there a particular commit related to the bugfix that you think needs review?  Bug squashing is not usually a contentious issue for me.  :stuck_out_tongue: 

I don't know the answer to your template question but I hope it works as simply as this.

Horneth commented:
Yes it's this one https://github.com/broadinstitute/cromwell/commit/50cf0bd4f4f89bac5a82fe888166b3203bd5e1b7

The way evaluation is done makes nested evaluation hard to handle. I'm wondering if some kind of "multi-pass evaluation" wouldn't make things easier. 
The problem is that when evaluating something like `read_string("${file_name}")`,
the evaluation of `"${file_name}"` gives you a string which is most likely the name of a local file in the call directory, something like `out`. Now it's impossible just with this to know if it's supposed to be a gcs path, or a local path, which makes it difficult to know which "IoInterface" to use to do the `read_string`.
I ended up overwriting `fileContentsToString` in the JesEngineFunctions to append the gcs path of the call bucket before the file name, and also not providing JesBackend workflows with a `SharedFilesystemIoInterface`, because `out` is a valid Unix path so it would try to use this interface on paths that are actually "relative gcs paths".

mcovarr commented on src/main/scala/cromwell/util/google/GoogleCredentialFactory.scala:
```diff
@@ -60,20 +63,25 @@ object GoogleCredentialFactory {
-      .build()
-      .setRefreshToken(token))
+  private def forClientSecrets(token: String): Option[Credential] = {
+    GoogleConf.userAuthMode map { userAuth =>
```
IntelliJ is warning that this should be simplified

Horneth replied:
Ha mine doesn't :s Probably a collect instead I guess ?

mcovarr replied:
it's saying to remove the `userAuth` and just do this as a partial function

mcovarr commented:
Sorry if this is a dumb question, but on a `SharedFilesystemBackend` that's using GCS inputs for a call, is it possible that `out` could be correctly interpreted as a "GCS relative path"?  i.e. wouldn't that have to be a local file?  Per the docs and my understanding of this code, this only localizes down GCS inputs and doesn't delocalize outputs, and it didn't seem there's the concept of a GCS scratch bucket that could serve as the basis of a relative path.

+1 to removing the possibility that JesBackend workflows could get their hands on a shared file system IO interface.  :smile:

Horneth commented:
Not a dumb question :smile: It shouldn't be a problem because `isValidPath` in the `GoogleCloudStorage` interface won't return true for `out` as a path, so the IoManager will skip `GcsInterface` and try `SharedInterface`.

However it's not great because when it's actually a "relative GCS path", it wouldn't pass `isValidPath` and we wouldn't even have chance to "correct" the path and make it absolute before trying to read. This is why if you look at the ioInterface that is assigned in case of a JesBackend in the `WorkflowDescriptor`, it's directly the `GcsInterface`, not an `IoManager`. This way we skip the `isValidPath` check.. It's looks a bit hacky but it some sense there's no reason to have a IoManager with only one interface, and like you said there's no reason to provide a SharedFS interface to a JesBackend.

geoffjentry commented on README.md:
```diff
@@ -911,16 +917,18 @@ It is also necessary to fill out the `google` stanza in the configuration file.
+  cromwellAuthenticationScheme = "service_account"
 
-  // If authScheme is "user"
+  // If cromwellAuthenticationScheme is "user_account"
```
apologies if I've already asked this at some point, but why the change? Isn't it implied that it's cromwell because it's in cromwell?

Horneth replied:
The idea was to make a clear distinction between the `userAuthenticationScheme` which helps authenticating a user (per workflow), and the `cromwellAuthenticationScheme` which is cromwell-wide.
But maybe `authenticationScheme` is clear enough ?

mcovarr replied:
I prefer this the way you have it to make clear there are two distinct types of authentication.

geoffjentry commented on src/main/scala/cromwell/binding/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -1,13 +1,23 @@
 
 trait WdlStandardLibraryFunctions extends WdlFunctions[WdlValue] {
-  private def fail(name: String) = Failure(new UnsupportedOperationException(s"$name() not implemented yet"))
+
```
unnecessary \n's

geoffjentry commented on src/main/scala/cromwell/binding/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -91,16 +101,55 @@ trait WdlStandardLibraryFunctions extends WdlFunctions[WdlValue] {
+    } yield file
+  }
+
+  protected def write_lines(params: Seq[Try[WdlValue]]): Try[WdlFile] = {
```
These can all be one line

geoffjentry commented on src/main/scala/cromwell/engine/io/IoManager.scala:
```diff
@@ -0,0 +1,31 @@
+  * Given a list of IoInterfaces, tries to find a suitable one to process the given path.
+  */
+class IoManager(interfaces: Seq[IoInterface]) extends IoInterface {
+
```
unnecessary \n

mcovarr commented on src/main/scala/cromwell/engine/backend/local/SharedFileSystem.scala:
```diff
@@ -129,8 +108,17 @@ trait SharedFileSystem {
   def postProcess(backendCall: LocalFileSystemBackendCall): Try[CallOutputs] = {
-    implicit val hasher = fileHasher(backendCall.workflowDescriptor)
+    //FIXME: Use the backend fileHasher
+    implicit val hasher = backendCall.workflowDescriptor.fileHasher
```
does this need FIXing?

Horneth replied:
Ah no, not anymore

mcovarr commented on src/main/scala/cromwell/engine/backend/local/SharedFileSystem.scala:
```diff
@@ -148,7 +136,7 @@ trait SharedFileSystem {
       Success(unwrappedMap.toMap)
     } else {
-      val message = taskOutputFailures collect { case (name, Failure(e)) => s"$name: $e" }
+      val message = taskOutputFailures collect { case (name, Failure(e)) => s"$name: $e\n${e.getStackTrace}" }
```
I think you want the `ExceptionUtils` version, this will produce gibberish
```
scala> val e = new Throwable
e: Throwable = java.lang.Throwable

scala> e.getStackTrace.toString
res2: String = [Ljava.lang.StackTraceElement;@63753b6d
```

mcovarr commented on src/main/scala/cromwell/engine/backend/sge/SgeBackendCall.scala:
```diff
@@ -19,7 +19,7 @@ case class SgeBackendCall(backend: SgeBackend,
   val script = callRootPath.resolve("script.sh")
   val returnCode = callRootPath.resolve("rc")
-  val engineFunctions: SgeEngineFunctions = new SgeEngineFunctions(callRootPath, stdout, stderr, workflowDescriptor.ioInterface)
+  val engineFunctions: SgeEngineFunctions = new SgeEngineFunctions(callRootPath, stdout, stderr, workflowDescriptor.ioManager)
```
why is the field of type `IoInterface` called `ioManager`?

Horneth replied:
`IoManager` is a bit strange in a sense that even though it `extends IoInterface` it doesn't really "implement" it, it uses actual `IoInterface` implementations (GCS, Shared FileSystem) and redirect to the correct one depending on the shape of the path.
The name `ioManager` was an attempt to capture this oddity, but I'm fine calling this an ioInterface if it's less confusing.

mcovarr commented on src/main/scala/cromwell/engine/io/gcs/GoogleCloudStorage.scala:
```diff
@@ -23,8 +23,26 @@ import scala.language.postfixOps
+
+  def userAuthenticated(workflowOptions: WorkflowOptions): Option[GoogleCloudStorage] = {
+    for {
+      conf <- GoogleConfiguration.gcloudConf.toOption
```
not sure of the interplay between this and `cromwellAuthenticated`, but if `gcloudConf` is a `Failure` we don't log the error anywhere along the path and then `toOption` it out of existence

Horneth replied:
We can make it something like this instead
```
def userAuthenticated(workflowOptions: WorkflowOptions): Option[GoogleCloudStorage] = {
    GoogleConfiguration.gcloudConf map { conf =>
      for {
        token <- workflowOptions.get(RefreshTokenOptionKey).toOption
        userCredentials <- GoogleCredentialFactory.fromUserAuthScheme(token)
      } yield apply(conf.appName, userCredentials)
    } get
  }
```
This way we'll fail if we can't find a google configuration (which we should anyway, with the error message being the validation error), and we still get an `Option[Credential]` depending on if the user configuration (clientId / Secret + refresh token) was provided.

mcovarr commented on src/test/scala/cromwell/engine/db/slick/SlickDataAccessSpec.scala:
```diff
@@ -81,13 +83,11 @@ class SlickDataAccessSpec extends FlatSpec with Matchers with ScalaFutures {
       throw new NotImplementedError
 
-    override def ioInterface(workflowOptions: WorkflowOptions): IOInterface = throw new NotImplementedError
+    override def workflowContext(workflowOptions: WorkflowOptions, workflowId: WorkflowId, name: String): WorkflowContext = ???
```
NotImplementedError

mcovarr commented on src/test/scala/cromwell/util/docker/SprayDockerRegistryApiClientSpec.scala:
```diff
@@ -130,8 +133,8 @@ with IntegrationPatience {
       val exception = client.getDockerHash(identifier).failed.futureValue
-      exception shouldBe a[Missing]
-      exception.getMessage should be("No configuration setting found for key 'google.authScheme'")
+      exception shouldBe a[UnsuccessfulResponseException]
```
super nitpick: should be `an`

geoffjentry replied:
![61wyynsmskl _sy344_bo1 204 203 200_](https://cloud.githubusercontent.com/assets/961771/11941573/9c494bdc-a7fe-11e5-9c75-dee169e0dcc1.jpg)

mcovarr replied:
yeah this nitpick rates at least a Michelin star

mcovarr commented:
:+1:
